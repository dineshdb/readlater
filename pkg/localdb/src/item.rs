use std::collections::HashSet;

use serde::Deserialize;
use serde::Serialize;

#[derive(Deserialize, Debug, Clone)]
pub struct Item {
    pub id: i64,
    pub pocket_id: Option<i64>,
    pub title: String,
    pub url: String,
    pub excerpt: Option<String>,

    /// Extended metadata about the url
    pub is_article: Option<bool>,
    pub is_index: Option<bool>,
    pub has_video: Option<HasVideo>,
    pub has_image: Option<HasImage>,
    pub word_count: Option<i32>,
    pub lang: Option<String>,
    pub time_to_read: Option<i32>,
    pub top_image_url: Option<String>,
    pub listen_duration_estimate: Option<i32>,
    pub tags: HashSet<Tag>,

    // fields related to the status of the url with respect to the user
    pub status: ItemStatus,
    pub time_added: i32,
    pub time_updated: Option<i32>,
    pub time_read: Option<i32>,
    pub time_favorited: Option<i32>,
}

impl From<&pocket::Item> for Item {
    fn from(value: &pocket::Item) -> Self {
        Self {
            id: value.item_id,
            pocket_id: Some(value.item_id),
            title: value.resolved_title.clone(),
            url: value.resolved_url.clone(),
            excerpt: Some(value.excerpt.clone()),
            is_article: Some(value.is_article),
            is_index: Some(value.is_article),
            has_video: Some(value.has_video.into()),
            has_image: Some(value.has_image.into()),
            word_count: Some(value.word_count),
            lang: Some(value.lang.clone()),
            time_to_read: Some(value.time_to_read),
            top_image_url: value.top_image_url.clone(),
            listen_duration_estimate: Some(value.listen_duration_estimate),
            tags: value
                .tags
                .keys()
                .map(|tag| Tag {
                    id: 0,
                    tag: tag.clone(),
                    name: None,
                })
                .collect(),
            status: value.status.into(),
            time_added: value.time_added,
            time_updated: Some(value.time_updated),
            time_read: Some(value.time_read),
            time_favorited: Some(value.time_favorited),
        }
    }
}

impl Default for Item {
    fn default() -> Self {
        Item {
            id: 0,
            pocket_id: None,
            title: "Example URL".to_string(),
            url: "http://example.com".to_string(),
            excerpt: None,
            status: ItemStatus::Unread,
            time_added: 0,
            time_updated: None,
            time_read: None,
            time_favorited: None,
            tags: HashSet::new(),

            is_article: None,
            is_index: None,
            has_video: None,
            has_image: None,
            word_count: None,
            lang: None,
            time_to_read: None,
            top_image_url: None,
            listen_duration_estimate: None,
        }
    }
}

#[derive(Deserialize, Serialize, Debug, sqlx::FromRow, Clone, PartialEq, Eq, Hash)]
pub struct Tag {
    pub id: i32,
    pub tag: String,
    pub name: Option<String>,
}

#[derive(Deserialize, Debug, sqlx::Type, PartialEq, Eq, Clone, Copy)]
#[repr(i32)]
pub enum ItemStatus {
    Unread = 0,
    Archived = 1,
    Deleted = 2,
}

impl From<pocket::item::ItemStatus> for ItemStatus {
    fn from(status: pocket::item::ItemStatus) -> Self {
        match status {
            pocket::item::ItemStatus::Unread => ItemStatus::Unread,
            pocket::item::ItemStatus::Archived => ItemStatus::Archived,
            pocket::item::ItemStatus::Deleted => ItemStatus::Deleted,
        }
    }
}

impl From<ItemStatus> for pocket::item::ItemStatus {
    fn from(status: ItemStatus) -> Self {
        match status {
            ItemStatus::Unread => pocket::item::ItemStatus::Unread,
            ItemStatus::Archived => pocket::item::ItemStatus::Archived,
            ItemStatus::Deleted => pocket::item::ItemStatus::Deleted,
        }
    }
}
#[derive(Deserialize, Debug, Clone, sqlx::Type, Copy)]
#[repr(i32)]
pub enum HasVideo {
    No = 0,
    Yes = 1,
    IsVideo = 2,
}

impl From<pocket::item::HasVideo> for HasVideo {
    fn from(has_video: pocket::item::HasVideo) -> Self {
        match has_video {
            pocket::item::HasVideo::No => HasVideo::No,
            pocket::item::HasVideo::Yes => HasVideo::Yes,
            pocket::item::HasVideo::IsVideo => HasVideo::IsVideo,
        }
    }
}

impl From<HasVideo> for pocket::item::HasVideo {
    fn from(has_video: HasVideo) -> Self {
        match has_video {
            HasVideo::No => pocket::item::HasVideo::No,
            HasVideo::Yes => pocket::item::HasVideo::Yes,
            HasVideo::IsVideo => pocket::item::HasVideo::IsVideo,
        }
    }
}

#[derive(Deserialize, Debug, Clone, sqlx::Type, Copy)]
#[repr(i32)]
pub enum HasImage {
    No = 0,
    Yes = 1,
    IsImage = 2,
}

impl From<pocket::item::HasImage> for HasImage {
    fn from(has_image: pocket::item::HasImage) -> Self {
        match has_image {
            pocket::item::HasImage::No => HasImage::No,
            pocket::item::HasImage::Yes => HasImage::Yes,
            pocket::item::HasImage::IsImage => HasImage::IsImage,
        }
    }
}
