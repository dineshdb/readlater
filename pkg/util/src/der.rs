use serde::{
    de::{self, Unexpected},
    Deserialize, Deserializer,
};
use std::str::FromStr;

pub fn bool_from_number<'de, D>(deserializer: D) -> Result<bool, D::Error>
where
    D: Deserializer<'de>,
{
    let value: i64 = Deserialize::deserialize(deserializer)?;
    match value {
        0 => Ok(false),
        1 => Ok(true),
        _ => Err(de::Error::invalid_value(
            Unexpected::Signed(value),
            &"expected truthy or falsy string",
        )),
    }
}

pub fn from_string<'de, D, T: FromStr>(deserializer: D) -> Result<T, D::Error>
where
    D: Deserializer<'de>,
    T::Err: std::fmt::Display,
{
    let value: &str = Deserialize::deserialize(deserializer)?;
    value.parse::<T>().map_err(de::Error::custom)
}

pub fn opt_from_string<'de, D, T: FromStr>(deserializer: D) -> Result<Option<T>, D::Error>
where
    D: Deserializer<'de>,
    T::Err: std::fmt::Display,
{
    let v = from_string(deserializer);
    match v {
        Ok(v) => Ok(Some(v)),
        Err(_) => Ok(None),
    }
}
