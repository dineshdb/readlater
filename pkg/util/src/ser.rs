use serde::Serializer;

pub fn serialize_option_bool_as_int<S>(
    value: &Option<bool>,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    match value {
        Some(true) => serializer.serialize_i32(1),
        Some(false) | None => serializer.serialize_i32(0),
    }
}

pub fn ser_opt_as_str<S>(value: &Option<i32>, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    match value {
        Some(v) => serializer.serialize_str(v.to_string().as_str()),
        None => serializer.serialize_str("null"),
    }
}

pub fn ser_as_str<S>(value: i32, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    serializer.serialize_str(&value.to_string())
}

pub fn serialize_vec_as_comma_separated<S>(vec: &[String], serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    let joined = vec.join(",");
    serializer.serialize_str(&joined)
}
