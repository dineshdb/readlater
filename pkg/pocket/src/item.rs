use serde::Deserialize;
use serde::Serialize;
use std::collections::HashMap;
use util::der::bool_from_string;
use util::der::from_string;

#[derive(Deserialize, Debug)]
pub struct Item {
    #[serde(deserialize_with = "from_string")]
    pub item_id: i64,
    #[serde(deserialize_with = "from_string")]
    pub resolved_id: i64,
    pub given_url: String,
    pub given_title: String,
    #[serde(deserialize_with = "bool_from_string")]
    pub favorite: bool,
    pub status: ItemStatus,
    #[serde(deserialize_with = "from_string")]
    pub time_added: i32,
    #[serde(deserialize_with = "from_string")]
    pub time_updated: i32,
    #[serde(deserialize_with = "from_string")]
    pub time_read: i32,
    #[serde(deserialize_with = "from_string")]
    pub time_favorited: i32,
    pub sort_id: i32,
    pub resolved_title: String,
    pub resolved_url: String,
    pub excerpt: String,
    #[serde(deserialize_with = "bool_from_string")]
    pub is_article: bool,
    #[serde(deserialize_with = "bool_from_string")]
    pub is_index: bool,
    pub has_video: HasVideo,
    pub has_image: HasImage,
    #[serde(deserialize_with = "from_string")]
    pub word_count: i32,
    pub lang: String,
    pub time_to_read: i32,
    pub top_image_url: Option<String>,
    pub listen_duration_estimate: i32,
    pub tags: HashMap<String, Tag>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Tag {
    pub tag: String,
    #[serde(deserialize_with = "from_string")]
    pub item_id: u64,
}

#[derive(Deserialize, Debug, Clone)]
#[repr(i32)]
pub enum ItemStatus {
    #[serde(rename = "0")]
    Unread = 0,
    #[serde(rename = "1")]
    Archived = 1,
    #[serde(rename = "2")]
    Deleted = 2,
}

#[derive(Deserialize, Debug, Clone)]
#[repr(i32)]
pub enum HasVideo {
    #[serde(rename = "0")]
    No = 0,
    #[serde(rename = "1")]
    Yes = 1,
    #[serde(rename = "2")]
    IsVideo = 2,
}

#[derive(Deserialize, Debug, Clone)]
#[repr(i32)]
pub enum HasImage {
    #[serde(rename = "0")]
    No = 0,
    #[serde(rename = "1")]
    Yes = 1,
    #[serde(rename = "2")]
    IsImage = 2,
}
