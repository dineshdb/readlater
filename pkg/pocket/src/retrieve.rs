use super::item::Item;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use util::der::bool_from_number;
use util::der::opt_from_string;
use util::ser::ser_opt_as_str;
use util::ser::serialize_option_bool_as_int;

#[derive(Serialize)]
pub enum State {
    #[serde(rename = "unread")]
    Unread,
    #[serde(rename = "archive")]
    Archive,
    #[serde(rename = "all")]
    All,
}

#[derive(Serialize)]
#[serde(rename_all = "lowercase")]
pub enum DetailType {
    Simple,
    Complete,
}

#[derive(Serialize)]
pub enum Tag {
    #[serde(rename = "_untagged_")]
    Untagged,
    #[serde(untagged)]
    Value(String),
}

#[derive(Serialize)]
#[serde(rename_all = "lowercase")]
pub enum ContentType {
    Article,
    Video,
    Image,
}

#[derive(Serialize)]
#[serde(rename_all = "lowercase")]
pub enum SortBy {
    Newest,
    Oldest,
    Title,
    Site,
}

#[derive(Deserialize, Debug)]
pub struct RetrieveResponse {
    pub status: i32,
    pub error: Option<String>,
    #[serde(deserialize_with = "bool_from_number")]
    pub complete: bool,
    pub since: i32,
    pub offset: Option<i32>,
    pub count: Option<i32>,
    #[serde(deserialize_with = "opt_from_string")]
    pub total: Option<i32>,
    pub list: BTreeMap<String, Item>,
}

impl RetrieveResponse {
    pub fn has_more(&self) -> crate::PocketResult<bool> {
        if let Some(total) = self.total {
            let count = self.count.unwrap_or(self.list.len() as i32);
            let offset = self.offset.unwrap_or(0);
            return Ok(total > count + offset);
        }

        Err(crate::PocketError::InvalidPagintionRequest)
    }
}

#[derive(Serialize)]
pub struct GetOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<State>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(serialize_with = "serialize_option_bool_as_int")]
    pub favorite: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tag: Option<Tag>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_type: Option<ContentType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sort: Option<SortBy>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "detailType")]
    pub detail_type: Option<DetailType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub search: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(serialize_with = "ser_opt_as_str")]
    pub since: Option<i32>,
    pub count: i32,
    pub total: IncludeTotal,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<i32>,
}

#[derive(Serialize)]
pub enum IncludeTotal {
    #[serde(rename = "1")]
    Include,
    #[serde(rename = "0")]
    Exclude,
}

impl Default for GetOptions {
    fn default() -> Self {
        Self {
            state: Some(State::Unread),
            favorite: None,
            tag: None,
            content_type: None,
            sort: Some(SortBy::Oldest),
            detail_type: Some(DetailType::Simple),
            search: None,
            domain: None,
            total: IncludeTotal::Include,
            since: None,
            // 30 is the maximum value permitted
            count: 30,
            offset: None,
        }
    }
}

impl GetOptions {
    pub fn for_pagination() -> Self {
        Self {
            total: IncludeTotal::Include,
            ..Default::default()
        }
    }

    pub fn count(&mut self, count: i32) -> &mut Self {
        self.count = count;
        self
    }

    pub fn offset(&mut self, offset: i32) -> &mut Self {
        self.offset = Some(offset);
        self
    }

    pub fn build(&mut self) -> GetOptions {
        let mut options = GetOptions::default();
        std::mem::swap(self, &mut options);
        options
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_get_options_default() {
        let options = GetOptions::default();
        insta::assert_json_snapshot!(options);
    }

    #[test]
    fn test_get_options() {
        let options = GetOptions {
            state: Some(State::Archive),
            favorite: Some(true),
            tag: Some(Tag::Value("linux".to_string())),
            content_type: Some(ContentType::Article),
            sort: Some(SortBy::Title),
            detail_type: Some(DetailType::Complete),
            search: Some("example".to_string()),
            domain: Some("example.org".to_string()),
            since: Some(1738297033),
            count: 10,
            total: IncludeTotal::Include,
            offset: Some(10),
        };
        insta::assert_json_snapshot!(options);
    }
}
