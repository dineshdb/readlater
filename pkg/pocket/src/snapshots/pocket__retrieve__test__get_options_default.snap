---
source: pkg/pocket/src/retrieve.rs
expression: options
---
{
  "state": "unread",
  "sort": "oldest",
  "count": 30,
  "total": "0"
}
